sdk/core/core/inc/az_json.h

```diff
@@ -182,81 +196,233 @@ AZ_NODISCARD az_result az_json_token_get_string(az_json_token const* token, az_s
+ *         - #AZ_OK if the az_json_builder is initialized successfully
  */
-AZ_NODISCARD AZ_INLINE az_result
-az_json_builder_init(az_json_builder* json_builder, az_span json_buffer)
+AZ_NODISCARD AZ_INLINE az_result az_json_builder_init(
+    az_json_builder* json_builder,
+    az_span destination_buffer,
+    az_json_builder_options const* options)
 {
   *json_builder
-      = (az_json_builder){ ._internal = { .json = json_buffer, .length = 0, .need_comma = false } };
+      = (az_json_builder){ ._internal = {
+                               .destination_buffer = destination_buffer,
+                               .bytes_written = 0,
+                               .need_comma = false,
+                               .token_kind = AZ_JSON_TOKEN_NONE,
+                               .bit_stack = { 0 },
+                               .options
+                               = options == NULL ? az_json_builder_options_default() : *options,
+                           } };
   return AZ_OK;
 }
 
-/*
- * @brief az_json_builder_span_get returns the az_span containing the final JSON object.
+/**
+ * @brief Returns the #az_span containing the JSON text written to the underlying buffer so far.
+ *
+ * @param[in] json_builder A pointer to an #az_json_builder instance wrapping the destination
+ * buffer.
+ *
+ * @note Do NOT modify or override the contents of the returned #az_span unless you are no longer
+ * building JSON text into it.
  *
- * @param json_builder A pointer to an az_json_builder instance wrapping the JSON buffer.
- * @return an az_span containing the final JSON object.
+ * @return An #az_span containing the JSON text built so far.
  */
-AZ_NODISCARD AZ_INLINE az_span az_json_builder_span_get(az_json_builder const* json_builder)
+AZ_NODISCARD AZ_INLINE az_span az_json_builder_get_json(az_json_builder const* json_builder)
 {
-  return az_span_slice(json_builder->_internal.json, 0, json_builder->_internal.length);
+  // TODO: Consider precondition validation of the state before returning.
+  return az_span_slice(
+      json_builder->_internal.destination_buffer, 0, json_builder->_internal.bytes_written);
 }
 
-/*
- * @brief az_json_builder_append_token appends an az_json_token to the JSON buffer.
+/**
+ * @brief Appends the UTF-8 text value (as a JSON string) into the buffer.
  *
- * @param json_builder A pointer to an az_json_builder instance containing the buffer to append the
- * token to.
- * @param token A pointer to the az_json_token to append.
- * @return AZ_OK if the token was appended successfully.<br>
- * AZ_ERROR_BUFFER_OVERFLOW if the buffer is too small.
+ * @param[in] json_builder A pointer to an #az_json_builder instance containing the buffer to append
+ * the string value to.
+ * @param[in] value The UTF-8 encoded value to be written as a JSON string. The value is escaped
+ * before writing.
+ *
+ * @remarks If \p value is #AZ_SPAN_NULL, the JSON `null` value is written, as if the
```

Comment:
Good point. I had considered this. This comes down to whether we want to consider `AZ_SPAN_NULL` to be special/different from an empty span/string.

For example, in .NET, when a user passes in a null `string` to a `WriteString` JSON function, we write the `null` literal. And we intentionally distinguish between an `empty string` and a `null string`.

We didn't distinguish between an `empty span` and a `null span` though because spans aren't synonymous with strings (and we have separate method overloads). **In C, is our `az_span` considered a surrogate for `string` and  then if so, will the user think that passing in a AZ_SPAN_NULL should imply `null` where as an empty span should be `""`?**

> Perhaps appending 'null' should be its own function.

Such a function already exists (`append_null`). I added this in the `append_string` in case the user is passing in AZ_SPAN_NULL as a "string" representation, where they didn't create or own the `az_span`. This avoids them having to do the null check to figure out which method to call.

Reply:
I think trying to append any az_span with size 0 would end up appending nothing to the buffer.
AZ_SPAN_NULL would be a size 0 az_span that would just have no impact to what was written

Reply:
The more I think about this, the more resolute I become: *_append_string should append a string; not a null. If I trim whitespace from a string and the string contains only whitespace, then I get a 0-size STRING; not a NULL. If I pass this STRING to _append_string, I should get "".

Reply:
> *_append_string should append a string; not a null. If I trim whitespace from a string and the string contains only whitespace, then I get a 0-size STRING; not a NULL. If I pass this STRING to _append_string, I should get "".

That is exactly what will happen today, as long as trim returns a 0-size STRING, not a NULL.

If anything returns a null string however, it could be problematic to treat that as a `""` in JSON as that is very different from `null`.